apps/wallet/src/ui/app/redux/slices/transaction-requests/index.ts

```diff
@@ -65,10 +69,131 @@ export const loadTransactionResponseMetadata = createAsyncThunk<
     }
 );
 
+export type CoinsMetaProps = {
+    amount: number;
+    coinType: string;
+    receiverAddress: string;
+};
+
+export type TxnMetaResponse = {
+    objectIDs: string[];
+    coins: CoinsMetaProps[];
+};
+
+const getEventsSummary = (
+    txEffects: TransactionEffects,
+    address: string
+): TxnMetaResponse => {
+    const events = txEffects?.events || [];
+
+    const coinsMeta = events
+        .map((event) => {
+            if (
+                !('coinBalanceChange' in event) ||
+                (event?.coinBalanceChange?.changeType !== 'Receive' &&
+                    event?.coinBalanceChange?.changeType !== 'Pay')
+            ) {
+                return null;
+            }
+            /// Combine all the coin balance changes from Pay and Receive events
+            /// A net positive amount means the user received coins
+            /// A net negative amount means the user sent coins
+            const { coinBalanceChange } = event;
+            const { coinType, amount, coinObjectId, owner } = coinBalanceChange;
+            const { AddressOwner } = owner as { AddressOwner: string };
+            const { ObjectOwner } = owner as { ObjectOwner: string };
+
+            if (ObjectOwner) {
+                return null;
+            }
+
+            return {
+                amount: amount,
+                coinType: coinType,
+                coinObjectId: coinObjectId,
+                receiverAddress: AddressOwner,
+            };
+        })
+        .filter(notEmpty);
+
+    const objectIDs: string[] = events
+
+        .map((event) => {
+            if (!('transferObject' in event)) {
+                return null;
+            }
+            const { transferObject } = event;
+            const { AddressOwner } = transferObject.recipient as {
+                AddressOwner: string;
+            };
+            if (AddressOwner !== address) {
+                return null;
+            }
+            return transferObject?.objectId;
+        })
+        .filter(notEmpty);
+
+    /// Group coins by receiverAddress
+    // sum coins by coinType for each receiverAddress
+    const meta = coinsMeta.reduce((acc, value, _) => {
+        return {
+            ...acc,
+            [`${value.receiverAddress}${value.coinType}`]: {
+                amount:
+                    value.amount +
```

Comment:
The value in the event is a number? Seems like we can have precision issues? cc @patrickkuo

Reply:
yes, this is a number base on the type.